apis/python/tools/cartographer

```diff
@@ -0,0 +1,346 @@
+    atlas_uri: str
+    verbose: bool
+
+    OBS_COLUMNS = [
+        "assay_ontology_term_id",
+        "cell_type_ontology_term_id",
+        "development_stage_ontology_term_id",
+        "disease_ontology_term_id",
+        "ethnicity_ontology_term_id",
+        "organism_ontology_term_id",
+        "sex_ontology_term_id",
+        "tissue_ontology_term_id",
+        "assay",
+        "cell_type",
+        "development_stage",
+        "disease",
+        "ethnicity",
+        "organism",
+        "sex",
+        "tissue",
+    ]
+    VAR_COLUMNS = ["feature_biotype", "feature_name", "feature_reference"]
+
+    # ----------------------------------------------------------------
+    def __init__(
+        self,
+        atlas_uri: str,
+        verbose: bool = False,
+    ):
+        self.ctx = self._create_tiledb_ctx()
+        self.atlas_uri = atlas_uri
+        self.verbose = verbose
+        self._allow_non_primary_data = False
+
+    # ----------------------------------------------------------------
+    def _create_tiledb_ctx(self) -> tiledb.Ctx:
+        return tiledb.Ctx(
+            {
+                "vfs.s3.region": os.environ.get("AWS_DEFAULT_REGION", "us-west-2"),
+                "py.init_buffer_bytes": 4 * 1024**3,  # per-column buffer size
+            }
+        )
+
+    # ----------------------------------------------------------------
+    def add_h5ad(self, input_dataset_id: Optional[str], input_h5ad_path) -> int:
+        """Add an H5AD to the atlas"""
+
+        soco = self._init_soco()
+        soma_name = self._get_soma_name(input_dataset_id, input_h5ad_path)
+        if soma_name in soco:
+            raise Exception(f"SOMA {soma_name} is already in SOMACollection {soco.uri}")
+
+        logging.info("Loading H5AD")
+        ann = anndata.read_h5ad(input_h5ad_path)
+
+        self._clean_and_add(ann, soma_name, soco)
+        return 0
+
+    # ----------------------------------------------------------------
+    def add_soma(self, input_dataset_id: Optional[str], input_soma_uri: str) -> int:
+        """Add a SOMA to the atlas"""
+
+        soco = self._init_soco()
+        soma_name = self._get_soma_name(input_dataset_id, input_soma_uri)
+        if soma_name in soco:
+            raise Exception(f"SOMA {soma_name} is already in SOMACollection {soco.uri}")
+
+        logging.info("Loading SOMA")
+        input_soma = SOMA(input_soma_uri)
+        ann = SOMAio.to_anndata(input_soma)
+
+        self._clean_and_add(ann, soma_name, soco)
+        return 0
+
+    # ----------------------------------------------------------------
+    def _init_soco(self) -> SOMACollection:
+        """
+        Makes sure the destination SOMACollection exists for first write.
+        """
+        soco = SOMACollection(
+            self.atlas_uri, name="atlas", ctx=self.ctx, verbose=self.verbose
+        )
+        if not soco.exists():  # Must be done first, to create the parent directory
+            soco.create_unless_exists()
```

Comment:
a more flexible API would be along the lines of os.mkdir, which allows:
* create if it doesn't exist, otherwise no error
* create if it does not exist, error out if it does
using the signature:    `os.mkdr(path, exists_ok: bool)`

So I'd suggest:   `soco.create(exists_ok: bool = False)`

Reply:
thanks @bkmartinjr ! I'll do this as a separate PR